PBblas/internal/BlockDimensions.ecl

```diff
@@ -0,0 +1,112 @@
+  SHARED single_partition_size := 100 : STORED('BlockDimensions_single_partition_size');
+  // Function to round up a dimension to the nearest number of nodes
+  SHARED dimension_t round_up_to_nodes(dimension_t d) := FUNCTION
+    dimension_t rounded := IF(d % nodes > 0, d  + (nodes - d % nodes), d);
+    return rounded; 
+  END;
+
+  SHARED cells := N * M; // Total number of cells in matrix
+  SHARED is_square := N = M; // Need special handling for exactly square matrix
+  SHARED min_rc := MIN([N, M]); // The smallest dimension of N or M
+  SHARED max_rc := MAX([N,M]); // The largest dimension of N or M
+  min_D :=  ROUNDUP(SQRT(cells / max_partition_size)); // Constraint 1
+  max_D := min_rc; // Constraint 2
+  // Unless min_D happens to be an even divisor of both N and M, we need to
+  // adjust it so that each partition will still be less than max_partition_size
+  // when we round up the row and column count of the partition.
+  // This is a complicated adjustment that is explained more fully in Jira ML-290.
+  // Maximum error ratio -- compensates for two iterations (see ML-290 for explanation)
+  mer := (M*N + min_D*N + min_D*M) / (M*N);
+  temp_D := __COMMON__(IF(N % min_D = 0 AND M % min_D = 0, min_D, 
+    ROUNDUP(SQRT((cells + min_D*mer*(M + N)) / max_partition_size))));
+  temp_D1 := round_up_to_nodes(temp_D); // Constraint 3
+  // If the round up results in tiny partitions, just use the original.  This is
+  // because the round up can dwarf the original calc (e.g. if temp_D is 2 and nodes is
+  // 100).
+  temp_D2 := __COMMON__(IF(cells / (temp_D1 * temp_D1) < single_partition_size, temp_D, temp_D1));
+  // Special case for square matrixes.  Should always use square partitioning,
+  // even if not all nodes are used.
+  // Note:  if temp_D1 > max_D, square partitioning will fail and will
+  //  resort to row or column.  Thus, with the exception of the special-case
+  //  above, we will enforce Constraint 3 unless it conflicts with C2.
+  SHARED dimension_t test_D := IF(temp_D2 > max_D AND is_square, max_D, temp_D2);
+  
+  // Internal function to calculate a vector block size within a specified limit
+  SHARED dimension_t calc_vector_block_size(dimension_t dlimit) := FUNCTION
+  	min_partitions := ROUNDUP(cells / max_partition_size);
+  	other_dim := cells / dlimit;
+    temp_bsize := IF(min_partitions % dlimit = 0, min_partitions, ROUNDUP((cells + other_dim*min_partitions)/max_partition_size));
+    test_bsize := round_up_to_nodes(temp_bsize);
+    dimension_t bsize := IF(test_bsize <= dlimit, test_bsize, dlimit);
+    return bsize;
+  END;
+  /**
+    * The type of partitioning employed: Types.PartitionType.square, .row, or .column or .single
+    */
+  EXPORT Partitioning := IF(cells < single_partition_size, p_types.single,
+    IF(test_D = 1, p_types.single, IF(test_D <= min_rc, p_types.square, IF(N < M, p_types.column, p_types.row))));
```

Comment:
MAP(...) is usually a better choice in this case given you are returning a scalar (enumeration) than a nested IF(...) function.